File: string-paradigm_1/src/main/java/com/jesperancinha/string/paradigm/performance1/ParadigmDependency1Impl.java
package com.jesperancinha.string.paradigm.performance1;

import java.util.List;

import com.steelzack.string.paradigm.api.Dependency;

public class ParadigmDependency1Impl extends Dependency {
	private StringWrapper dependencies;

	public ParadigmDependency1Impl(StringWrapper data) {
		this.dependencies = data;
	}

	public int getDependencyCount() {
		return dependencies.toString().split(";").length;
	}

	public String getSubDependencyElement(int index) {
		final String[] splittedValues = dependencies.toString().split(";");
		if (index < splittedValues.length - 1) {
			return splittedValues[index];
		} else {
			return null;
		}
	}

	public boolean startsWith(List<String> subDependencyElement) {
		int dependencyCount = getDependencyCount();
		if (dependencyCount < subDependencyElement.size()) {
			return false;
		}
		for (int i = 0; i < subDependencyElement.size(); i++) {
			if (!getSubDependencyElement(i).equals(subDependencyElement.get(i).toString())) {
				return false;
			}
		}
		return true;
	}
}

File: string-paradigm_1/src/main/kotlin/org/jesperancinha/string/paradigm/performance1/ParadigmDependency1Impl.kt
package org.jesperancinha.string.paradigm.performance1

import org.jesperancinha.string.paradigm.api.Dependency

class ParadigmDependency1Impl(private val dependencies: StringWrapper) : Dependency() {
    override fun dependencyCount(): Int = dependencies.toString().split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray().size

    override fun getSubDependencyElement(index: Int): String? {
        val splittedValues = dependencies.toString().split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        return if (index < splittedValues.size - 1) {
            splittedValues[index]
        } else {
            null
        }
    }

    override fun startsWith(subDependencyElement: List<String>): Boolean {
        val dependencyCount = dependencyCount()
        if (dependencyCount < subDependencyElement.size) {
            return false
        }
        for (i in subDependencyElement.indices) {
            if (getSubDependencyElement(i) != subDependencyElement[i]) {
                return false
            }
        }
        return true
    }
}